Add docker-compose start/stop tasks

michael-yx-wu opened the pull request.

michael-yx-wu committed:
Add docker-compose start/stop tasks

uschi2000 commented:
So the only benefit of this thing is the ability to depend on the docker build task?

michael-yx-wu commented:
You can optionally use the `generateDockerCompose` task. You can also directly just call `dockerComposeUp` if you don't need it. It doesn't depend on a docker image being built by the `docker` plugin in this repo, if that's what you meant?

markelliot commented:
I think there's not quite enough configuration here for dockerComposeUp to be valuable over just calling `docker-compose up`.

markelliot commented on src/test/groovy/com/palantir/gradle/docker/DockerComposePluginTests.groovy:
```diff
@@ -130,4 +130,32 @@ class DockerComposePluginTests extends AbstractPluginTest {
+        buildResult.task(':dockerComposeUp').outcome == TaskOutcome.SUCCESS
+        buildResult.task(':dockerComposeStop').outcome == TaskOutcome.SUCCESS
+
+        offline.task(':dockerComposeRemove').outcome == TaskOutcome.SUCCESS
```
I think we need to find a way to test that the containers actually came up, then stopped, then removed in this test

michael-yx-wu commented:
Ashwin suggested adding the docker compose tasks in compass to the plugin to avoid duplication of such tasks across projects. Should I make it easier for users to add additional flags to docker-compose up?

> On Jan 25, 2016, at 10:35 PM, Mark Elliot notifications@github.com wrote:
> 
> I think there's not quite enough configuration here for dockerComposeUp to be valuable over just calling docker-compose up.
> 
> —
> Reply to this email directly or view it on GitHub.

markelliot commented:
Interesting. My concern is that we have one product now that looks like this for setup:

```
docker-compose -f docker-compose.yml --x-networking -p testnetwork up –d
```

And there's not really a way to provide the networking option. It seems like a bit of a pit of features to keep pace with how docker-compose sends arguments (the docker build commands are incredibly stable, which is why I opted to shell out in the other plugins here).

uschi2000 commented:
Sharing Mark's concerns; task dependency only doesn't justify the feature skin-hole that this will become.

michael-yx-wu committed:
add dockerComposeUpStatus task

Check whether the services specified in your docker-compose.yml file
are still running.

michael-yx-wu commented:
We could add a configuration block that allows users to specify additional arguments as well as whether the arguments apply to docker-compose or to a particular docker-compose command. It's unlikely people will want to add arguments to stop/remove. So we could do something like:

```
dockerCompose {
    composeFlags '--networking -p my-project'
    upFlags '--force-recreate'
}
```

markelliot commented:
I'm just worried about what it's going to take to capture the full set of arguments in a reasonable way — docker-compose is still under pretty heavy churn in terms of CLI API, I believe.

michael-yx-wu commented:
Okay, that makes sense. I'm closing this pull request for now then. We can revisit when we feel that the cli docker-compose is less volatile.
